File: OpenIM-SDK/src/main/java/io/openim/android/sdk/listener/OnAdvanceMsgListener.java
package io.openim.android.sdk.listener;

import java.util.List;

import io.openim.android.sdk.models.C2CReadReceiptInfo;
import io.openim.android.sdk.models.GroupMessageReceipt;
import io.openim.android.sdk.models.KeyValue;
import io.openim.android.sdk.models.Message;
import io.openim.android.sdk.models.RevokedInfo;

/**
 * Message listener
 */
public interface OnAdvanceMsgListener {
    /**
     * New message received; add to list and refresh UI
     */
    default void onRecvNewMessage(Message msg){}

    /**
     * Peer read receipt; update UI read status
     */
    default void onRecvC2CReadReceipt(List<C2CReadReceiptInfo> list){}

    /**
     * Group member read receipt; update UI read status
     */
    default void onRecvGroupMessageReadReceipt(GroupMessageReceipt groupMessageReceipt){}

    /**
     * Message recalled by peer. Includes single chats, group chats, and admin recalls.
     * The new version only returns recall details via this callback and does not trigger onRecvNewMessage.
     */
    default void onRecvMessageRevokedV2(RevokedInfo info){}

    default void onRecvMessageExtensionsChanged(String msgID, List<KeyValue> list){}

    default void onRecvMessageExtensionsDeleted(String msgID, List<String> list){}

    default void onRecvMessageExtensionsAdded(String msgID, List<KeyValue> list){}

    /**
     * Message deleted
     * @param message Deleted message
     */
    default void onMsgDeleted(Message message){}

    default void onRecvOfflineNewMessage(List<Message> msg){}

    default void onRecvOnlineOnlyMessage(String s){};
}

File: OpenIM-SDK/src/main/java/io/openim/android/sdk/listener/OnAdvanceMsgListener.kt
package io.openim.android.sdk.listener

import io.openim.android.sdk.models.C2CReadReceiptInfo
import io.openim.android.sdk.models.GroupMessageReceipt
import io.openim.android.sdk.models.KeyValue
import io.openim.android.sdk.models.Message
import io.openim.android.sdk.models.RevokedInfo

interface OnAdvanceMsgListener {
    fun onRecvNewMessage(msg: Message) {}
    fun onRecvC2CReadReceipt(list: List<C2CReadReceiptInfo>) {}
    fun onRecvGroupMessageReadReceipt(groupMessageReceipt: GroupMessageReceipt) {}
    fun onRecvMessageRevokedV2(info: RevokedInfo) {}
    fun onRecvMessageExtensionsChanged(msgID: String, list: List<KeyValue>) {}
    fun onRecvMessageExtensionsDeleted(msgID: String, list: List<String>) {}
    fun onRecvMessageExtensionsAdded(msgID: String, list: List<KeyValue>) {}
    fun onMsgDeleted(message: Message) {}
    fun onRecvOfflineNewMessage(msg: List<Message>) {}
    fun onRecvOnlineOnlyMessage(s: String) {}
}
